src/invite-api.js

```diff
@@ -0,0 +1,199 @@
+// @ts-check
+import { TypedEmitter } from 'tiny-typed-emitter'
+import { InviteResponse_Decision } from './generated/rpc.js'
+import { projectKeyToId } from './utils.js'
+
+/**
+ * @typedef {Object} InviteApiEvents
+ *
+ * @property {(info: { projectId: string, projectName?: string, peerId: string }) => void} invite-received
+ */
+
+/**
+ * @template K
+ * @template V
+ * @extends {Map<K, Set<V>>}
+ */
+class MapOfSets extends Map {
+  /** @param {K} key */
+  get(key) {
+    const existing = super.get(key)
+    if (existing) return existing
+    const set = new Set()
+    super.set(key, set)
+    return set
+  }
+}
+
+/**
+ * @extends {TypedEmitter<InviteApiEvents>}
+ */
+export class InviteApi extends TypedEmitter {
+  // Maps project id -> set of device ids
+  /** @type {MapOfSets<string, string>} */
+  #peersToRespondTo = new MapOfSets()
+
+  // Maps project id -> { invite, fromPeerId }
+  /** @type {Map<string, { fromPeerId: string, invite: import('./generated/rpc.js').Invite }>} */
+  #pendingInvites = new Map()
+
+  #isMember
+  #addProject
+
+  /**
+   * @param {Object} options
+   * @param {import('./rpc/index.js').MapeoRPC} options.rpc
+   * @param {object} options.queries
+   * @param {(projectId: string) => Promise<boolean>} options.queries.isMember
+   * @param {(invite: import('./generated/rpc.js').Invite) => Promise<void>} options.queries.addProject
+   */
+  constructor({ rpc, queries }) {
+    super()
+    this.rpc = rpc
+    this.#isMember = queries.isMember
+    this.#addProject = queries.addProject
+
+    this.rpc.on('invite', (peerId, invite) => {
+      this.#handleInvite(peerId, invite).catch(() => {
+        /* c8 ignore next */
+        // TODO: Log errors, but otherwise ignore them, but can't think of a reason there would be an error here
+      })
+    })
+  }
+
+  /**
+   * @param {string} peerId
+   * @param {import('./generated/rpc.js').Invite} invite
+   */
+  async #handleInvite(peerId, invite) {
+    const projectId = projectKeyToId(invite.projectKey)
+    const isAlreadyMember = await this.#isMember(projectId)
+
+    if (isAlreadyMember) {
+      this.#sendAlreadyResponse({ peerId, projectId })
+      return
+    }
+
+    const peerIds = this.#peersToRespondTo.get(projectId)
+    peerIds.add(peerId)
+
+    if (this.#pendingInvites.has(projectId)) return
+
+    this.#pendingInvites.set(projectId, { fromPeerId: peerId, invite })
+    this.emit('invite-received', {
+      // TODO: Should this be the project public ID since it can be exposed to the client?
+      // Probably would require changing the public methods to accept the public ID
+      // and using the public ID for #invites and #peersToRespondTo keys instead
+      peerId,
+      projectId,
+      projectName: invite.projectInfo?.name,
+    })
```

Comment:
@gmaclennan any thoughts on the question posed here? my understanding is that whatever's emitted here can be surfaced to a client UI. If that's the case, would it be preferable to expose "public" ids instead of or in addition to what's here, or do we let the class consumer (or end-user client) handle that?

Reply:
looks like we want this to be the public id so that it can be directly used in other project-related APIs (which accept a public id)